### ISSUE TITLE
TD File Extension Convention

### ISSUE THREAD

**Issue by egekorkan** [w3c/wot-thing-description#1214]
I had thought that we agreed at some point that the file extension for TDs should be `td.jsonld` or `td.json`. However, it seems that the IANA registration used `.jsontd` (https://www.iana.org/assignments/media-types/application/td+json) that now even Firefox downloads a TD with `.jsontd` extension. Should we change this convention we agreed or change IANA registration somehow and hope that it is not too late?

**Comment by relu91** [comment]
I don't know how much an effort is to change the IANA registration, but I want to stress that having `td.json` and `td.jsonld` as extensions would allow IDEs and browsers to fallback to JSON if they don't support `td` types. It is also al well establish a pattern in the `Nodejs` community and not only there.

**Comment by benfrancis** [comment]
As a data point, a similar W3C specification I've worked on, [Web App Manifest](https://www.w3.org/TR/wot-thing-description/) ([application/manifest+json](https://www.iana.org/assignments/media-types/application/manifest+json)) has a file extension of `.webmanifest`, but the specification says in a [note](https://www.w3.org/TR/appmanifest/#example-2-linking-to-a-manifest) that `.json` is also fine. In my experience the vast majority of developers use `.json`.

In that case the important point from the point of view of implementations is the `rel=manifest` link relation type which is used to discover manifests, and the MIME type is seen as more important than the file extension (which can also be omitted entirely).

I don't have a strong opinion about whether to try to change the IANA registration for Thing Descriptions, but if that turns out to be difficult one option would be to add a note to the TD specification to say that a `.json` file extension is also acceptable. The most important point is probably that the WoT Discovery specification doesn't depend on the file extension for discovery.

I agree that for developer tools having a `.json` file extension would be convenient as more tools will know how to handle the file.

**Comment by egekorkan** [comment]
Thanks for the comments @relu91 and @benfrancis . I think that the minimum we need to do is to explain that as a note regarding what is registered in IANA and what we propose. Ideally, we should change IANA registration since changing it in various clients that can download a TD is probably equally/more difficult and technically wrong.

On the other hand, I have looked at other `+` using MIME types at IANA and they either use just `.json` or `.xml` or define something new like we had done at the time with `.jsontd`. Could it be that a two-dotted file extension is not acceptable by IANA?

My opinion is to also use `.td.json`

**Comment by FadySalama** [comment]
I agree with @egekorkan that `.td.json` is a much better option than `.jsontd`. But I have no idea how things work at IANA. Maybe we can ask Matthias Kovatsch as it seems that he was responsible for the registration.

**Comment by sebastiankb** [comment]
As I understand it, only the last dot in the file name indicates the file type by the OS. So, it makes sense to have `.jsontd` as file extension as unique TD file type identifier. I would also support the idea from @benfrancis to simple provide a note that other convention are possible since the TD is json based. 

Btw: Changing the IANA entry will not help since the IANA registry is also part of the [W3C 1.0 Spec](https://www.w3.org/TR/wot-thing-description/#media-type-section).

**Comment by egekorkan** [comment]
Regarding recognizing the file, it depends on the IDE or Editor, the last dot is not necessarily the only way. Modern IDEs can be configured to detect myFile.td.json as a TD. This is what the extension developed by @relu91 for VS Code does (https://marketplace.visualstudio.com/items?itemName=arces-wot.td-code). Also, VS Code detects babel.config.js as a configuration file for babel and finds the correct icon. 

So I would say that we do not have constrain ourselves to having a single part after the dot

**Comment by sebastiankb** [comment]
from today's call:
- there is another proposal to extend the IANA registry with two additional file extensions ".td.json" and ".td.jsonld"
- we should talk with @plehegar about this issue
- we should also involve @mkovatsc since he did the registration
- we should also clearify if we need also a new IANA entry for the Thing Models, which should then use the same pattern as the TD file extensions

**Comment by sebastiankb** [comment]
from today's TD call:
- @sebastiankb will ping @mkovatsc and ask him to update the variations of the file extensions (see comment above)

**Comment by mkovatsc** [comment]
I will ask for some IANA advice how to handle this at next week's IETF meeting.

**Comment by mkovatsc** [comment]
A short summary of the big picture would be good:

* Do you want a separate media type for TMs?
  * Do you want to add `application/tm+json`?
  * Do you want to introduce parameters, e.g., `.../td+json;model` (I recall we can fully define the structure, so not necessary to have key=value)
* Could a single media type work in practice?
  * Parameters would belong to this option...
  * Is the evaluation of the `@type` in the document maybe enough to differentiate TMs and TDs?
  * If no need for a new media type, no need for a new file extension nor IANA change...
* Other points?

**Comment by sebastiankb** [comment]
@mkovatsc many thanks for taking this point

Actually, we have already decided for a separated registration for TM, which is currently under review: https://tools.iana.org/public-view/viewticket/1233681

>Do you want to introduce parameters, e.g., .../td+json;model (I recall we can fully define the structure, so not necessary to have key=value)

One disadvantage do I see is that the media type becomes quite complex ("if-then" statements are required). Another disadvantage could be (and what I sometimes discover in some tools) is that you can only select the media type without adding parameters. 

But maybe that's a good practice with the parameter approach. Do you maybe have a good example that we can look at?
